Get data from github

In [ ]:
#!wget https://raw.githubusercontent.com/shitkov/bert4classification/main/train.csv
#!wget https://raw.githubusercontent.com/shitkov/bert4classification/main/valid.csv
#!wget https://raw.githubusercontent.com/shitkov/bert4classification/main/test.csv

#!wget https://github.com/ArtyomR/AI-News-Classification/blob/main/ai_articles_test_231124.xlsx
#!wget https://github.com/ArtyomR/AI-News-Classification/blob/main/ai_articles_train_231124.xlsx
!wget https://raw.githubusercontent.com/shitkov/bert4classification/main/bert_classifier.py
!wget https://raw.githubusercontent.com/shitkov/bert4classification/main/bert_dataset.py
!wget https://raw.githubusercontent.com/shitkov/bert4classification/main/requirements.txt

--2023-12-21 12:32:23--  https://raw.githubusercontent.com/shitkov/bert4classification/main/bert_classifier.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.108.133, 185.199.109.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.108.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 5278 (5.2K) [text/plain]
Saving to: ‘bert_classifier.py’

bert_classifier.py  100%[===================>]   5.15K  --.-KB/s    in 0s      

2023-12-21 12:32:23 (42.0 MB/s) - ‘bert_classifier.py’ saved [5278/5278]

--2023-12-21 12:32:23--  https://raw.githubusercontent.com/shitkov/bert4classification/main/bert_dataset.py
Resolving raw.githubusercontent.com (raw.githubusercontent.com)... 185.199.109.133, 185.199.108.133, 185.199.110.133, ...
Connecting to raw.githubusercontent.com (raw.githubusercontent.com)|185.199.109.133|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 891 [tex

Installing Packages

In [ ]:
!pip install -r /content/requirements.txt

ERROR: Could not find a version that satisfies the requirement torch==1.8.1 (from versions: 1.11.0, 1.12.0, 1.12.1, 1.13.0, 1.13.1, 2.0.0, 2.0.1, 2.1.0, 2.1.1, 2.1.2)
ERROR: No matching distribution found for torch==1.8.1


In [ ]:
import pandas as pd
from sklearn.model_selection import train_test_split

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
train_valid_data = pd.read_excel('/content/drive/MyDrive/Data2/ai_articles_train_231124.xlsx', index_col=0)
train_valid_data.head(2)

published_flag                                   combined_text_ru  \
723                1  Школьный гимн и музыка на электрогитаре, котор...   
2997               0  Недавний минимум в 185 юаней: самостоятельная ...   

                                 combined_text_ru_clean  
723   школьный гимн и музыка на электрогитара которы...  
2997  недавний минимум в 185 юань самостоятельный по...

In [ ]:
test_data  = pd.read_excel('/content/drive/MyDrive/Data2/ai_articles_test_231124.xlsx',index_col=0)
test_data.head(2)

published_flag                                   combined_text_ru  \
1167               1  Tencent Cloud запускает продукт для рисования ...   
1542               1  Генеративный ИИ достиг «пика завышенных ожидан...   

                                 combined_text_ru_clean  
1167  tencent cloud запускать продукт для рисование ...  
1542  генеративный ия достигнуть пик завысить ожидан...

In [ ]:
#train_data = pd.read_csv('/content/train.csv')
"""
try:
    train_valid_data = pd.read_excel('/content/ai_articles_train_231124.xlsx', engine='openpyxl')
except ValueError:
    train_valid_data = pd.read_excel('/content/ai_articles_train_231124.xlsx', engine='xlrd')

#train_valid_data=pd.read_excel('/content/ai_articles_train_231124.xlsx')
#valid_data = pd.read_csv('/content/valid.csv')
#test_data  = pd.read_csv('/content/test.csv')
test_data  = pd.read_excel('/content/ai_articles_test_231124.xlsx')
"""

"\ntry:\n    train_valid_data = pd.read_excel('/content/ai_articles_train_231124.xlsx', engine='openpyxl')\nexcept ValueError:\n    train_valid_data = pd.read_excel('/content/ai_articles_train_231124.xlsx', engine='xlrd')\n\n#train_valid_data=pd.read_excel('/content/ai_articles_train_231124.xlsx')\n#valid_data = pd.read_csv('/content/valid.csv')\n#test_data  = pd.read_csv('/content/test.csv')\ntest_data  = pd.read_excel('/content/ai_articles_test_231124.xlsx')\n"

In [ ]:
from bert_dataset import CustomDataset
from bert_classifier import BertClassifier

Initialize BERT classifier

In [ ]:
classifier = BertClassifier(
        model_path='cointegrated/rubert-tiny2',
        tokenizer_path='cointegrated/rubert-tiny2',
        n_classes=2,
        epochs=10,
        model_save_path='/content/bert.pt'
)

config.json:   0%|          | 0.00/693 [00:00<?, ?B/s]

model.safetensors:   0%|          | 0.00/118M [00:00<?, ?B/s]

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at cointegrated/rubert-tiny2 and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


tokenizer_config.json:   0%|          | 0.00/401 [00:00<?, ?B/s]

vocab.txt:   0%|          | 0.00/1.08M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/112 [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.74M [00:00<?, ?B/s]

Prepare data and helpers for train and evlauation

In [ ]:
# split dtaset on train and valid_data dataframes
train_data, valid_data = train_test_split(
    train_valid_data,
    test_size=0.2,
    random_state=72
)
print(len(train_data),train_data.head(2),len(valid_data))

1726       published_flag                                   combined_text_ru  \
603                1  Китай активизирует усилия по созданию чипов ис...   
2621               1  Первая партия Xiaoice AI-клонов интернет-знаме...   

                                 combined_text_ru_clean  
603   китай активизировать усилие по создание чип ис...  
2621  первый партия xiaoice aiклон интернетзнаменито...   432


In [ ]:
classifier.preparation(
        X_train=list(train_data['combined_text_ru']),
        y_train=list(train_data['published_flag']),
        X_valid=list(valid_data['combined_text_ru']),
        y_valid=list(valid_data['published_flag'])
    )

/usr/local/lib/python3.10/dist-packages/transformers/optimization.py:411: FutureWarning: This implementation of AdamW is deprecated and will be removed in a future version. Use the PyTorch implementation torch.optim.AdamW instead, or set `no_deprecation_warning=True` to disable this warning
  warnings.warn(


Train loop

In [ ]:
%%time
classifier.train()

Epoch 1/10
Train loss 0.6683222349098563 accuracy 0.7410196987253765
Val loss 0.8898354326288596 accuracy 0.7754629629629629
----------
Epoch 2/10
Train loss 0.601306774071148 accuracy 0.8290845886442642
Val loss 1.0114193948331243 accuracy 0.7847222222222222
----------
Epoch 3/10
Train loss 0.4773395611666605 accuracy 0.8823870220162224
Val loss 1.0676517093921172 accuracy 0.7777777777777778
----------
Epoch 4/10
Train loss 0.35288639233754937 accuracy 0.9165701042873696
Val loss 1.1668002901584529 accuracy 0.7777777777777778
----------
Epoch 5/10
Train loss 0.27209732885886306 accuracy 0.9403244495944381
Val loss 1.297795055209193 accuracy 0.7800925925925926
----------
Epoch 6/10
Train loss 0.1889517457665758 accuracy 0.958864426419467
Val loss 1.3653970416127643 accuracy 0.7847222222222222
----------
Epoch 7/10
Train loss 0.1416633172459709 accuracy 0.9698725376593279
Val loss 1.4841758518465482 accuracy 0.7777777777777778
----------
Epoch 8/10
Train loss 0.11624129280742083 accurac

Check test data

In [ ]:
%%time
texts = list(test_data['combined_text_ru'])
labels = list(test_data['published_flag'])

predictions = [classifier.predict(t) for t in texts]

In [ ]:
from sklearn.metrics import precision_recall_fscore_support

precision, recall, f1score = precision_recall_fscore_support(labels, predictions,average='macro')[:3]

print(f'precision: {precision}, recall: {recall}, f1score: {f1score}')

precision: 0.8163239436619718, recall: 0.8200427502671892, f1score: 0.8175692334916094


In [ ]:
from sklearn.metrics import accuracy_score, classification_report

In [ ]:
# Model accuracy
accuracy = accuracy_score(labels, predictions)
print(f'Accuracy: {accuracy}')

# Additional metrics
print(classification_report(labels, predictions))

Accuracy: 0.8196544276457883
              precision    recall  f1-score   support

           0       0.86      0.82      0.84       525
           1       0.77      0.82      0.80       401

    accuracy                           0.82       926
   macro avg       0.82      0.82      0.82       926
weighted avg       0.82      0.82      0.82       926



**10 epochs**  
Accuracy: 0.8196544276457883  
              precision    recall  f1-score   support

           0       0.84      0.85      0.84       525
           1       0.80      0.78      0.79       401

    accuracy                           0.82       926
   macro avg       0.82      0.82      0.82       926  
weighted avg       0.82      0.82      0.82       926


**2 epochs**  
Accuracy: 0.7980561555075594  
    precision    recall  f1-score   support

           0       0.83      0.81      0.82       525
           1       0.76      0.79      0.77       401

    accuracy                           0.80       926
   macro avg       0.79      0.80      0.80       926  
weighted avg       0.80      0.80      0.80       926





## References
- [BERT для классификации русскоязычных текстов](https://habr.com/ru/articles/567028/)
- [News Article Classification Task using SOTA models and their comparison](https://medium.com/mlearning-ai/news-article-classification-task-using-sota-models-and-their-comparison-a0f87abdc0c1)